tools/state-viewer/src/state_parts.rs

```diff
@@ -0,0 +1,432 @@
+    let mut cur_block_info = epoch_manager.get_block_info(&head.last_block_hash).unwrap();
+    // EpochManager doesn't have an API that maps EpochId to Blocks, and this function works
+    // around that limitation by iterating over the epochs.
+    // This workaround is acceptable because:
+    // 1) Extending EpochManager's API is a major change.
+    // 2) This use case is not critical at all.
+    loop {
+        let cur_epoch_info = epoch_manager.get_epoch_info(cur_block_info.epoch_id()).unwrap();
+        let cur_epoch_height = cur_epoch_info.epoch_height();
+        assert!(
+            cur_epoch_height >= epoch_info.epoch_height(),
+            "cur_block_info: {:#?}, epoch_info.epoch_height: {}",
+            cur_block_info,
+            epoch_info.epoch_height()
+        );
+        let epoch_first_block_info =
+            epoch_manager.get_block_info(cur_block_info.epoch_first_block()).unwrap();
+        let prev_epoch_last_block_info =
+            epoch_manager.get_block_info(epoch_first_block_info.prev_hash()).unwrap();
+
+        if cur_epoch_height == epoch_info.epoch_height() {
+            return *prev_epoch_last_block_info.hash();
+        }
+
+        cur_block_info = prev_epoch_last_block_info;
+    }
+}
+
+pub(crate) fn apply_state_parts(
+    epoch_selection: EpochSelection,
+    shard_id: ShardId,
+    part_id: Option<u64>,
+    home_dir: &Path,
+    near_config: NearConfig,
+    store: Store,
+    root_dir: Option<PathBuf>,
+    s3_bucket: Option<String>,
+    s3_region: Option<String>,
+) {
+    let runtime_adapter: Arc<dyn RuntimeAdapter> =
+        Arc::new(NightshadeRuntime::from_config(home_dir, store.clone(), &near_config));
+    let mut epoch_manager =
+        EpochManager::new_from_genesis_config(store.clone(), &near_config.genesis.config)
+            .expect("Failed to start Epoch Manager");
+    let mut chain_store = ChainStore::new(
+        store.clone(),
+        near_config.genesis.config.genesis_height,
+        near_config.client_config.save_trie_changes,
+    );
+
+    let epoch_id = epoch_selection.to_epoch_id(store, &mut chain_store, &mut epoch_manager);
+    let epoch = runtime_adapter.get_epoch_info(&epoch_id).unwrap();
+    let sync_prev_hash = get_prev_hash_of_epoch(&epoch, &mut chain_store, &mut epoch_manager);
+    let sync_prev_block = chain_store.get_block(&sync_prev_hash).unwrap();
+
+    assert!(runtime_adapter.is_next_block_epoch_start(&sync_prev_hash).unwrap());
+    assert!(
+        shard_id < sync_prev_block.chunks().len() as u64,
+        "shard_id: {}, #shards: {}",
+        shard_id,
+        sync_prev_block.chunks().len()
+    );
+    let state_root = sync_prev_block.chunks()[shard_id as usize].prev_state_root();
+
+    let part_storage: Box<dyn StatePartReader> = if let Some(root_dir) = root_dir {
+        if s3_bucket.is_some() || s3_region.is_some() {
+            panic!("Output can only be done to the local filesystem or to S3, not both");
+        }
```

Comment:
Could this be moved to some clap parsing validation logic?

Reply:
Would be happy to, but don't find such support in `clap`.